p9/messages.go

```diff
@@ -1999,6 +2000,192 @@ func (r *rusymlink) String() string {
+// };
+// flags bits are:
+
+// #define P9_LOCK_FLAGS_BLOCK 1    /* blocking request */
+// #define P9_LOCK_FLAGS_RECLAIM 2  /* reserved for future use */
+
+// LockStatus contains lock status result.
+type LockStatus uint8
+
+// These are the four current return values for Rlock.
+const (
+	lockOK LockStatus = iota
+	lockBlocked
+	lockError
+	lockGrace
+)
+
+// ELockError is the overarching error for Tlock.
+// The 9P2000.L lock errors are basically "something failed",
+// i.e., not conformant with how 9P2000 is supposed to work.
+var ELockError = syscall.Errno(lockError)
+
+// The Linux v9fs client implements the fcntl(F_SETLKW) (blocking)
+// lock request by calling lock with P9_LOCK_FLAGS_BLOCK set. If the
+// response is P9_LOCK_BLOCKED, it retries the lock request in an
+// interruptible loop until status is no longer P9_LOCK_BLOCKED.
+
+// The Linux v9fs client translates BSD advisory locks (flock) to
+// whole-file POSIX record locks. v9fs does not implement mandatory
+// locks and will return ENOLCK if use is attempted.
+
+// Because of POSIX record lock inheritance and upgrade properties,
+// pass-through servers must be implemented carefully.
+
+// tlock is a Tlock message
+type tlock struct {
+	// fid is the fid to lock.
+	fid fid
+
+	Type   LockType // Type of lock: F_RDLCK, F_WRLCK, F_UNLCK */
+	Flags  uint32   // flags, not whence, docs are wrong.
+	Start  uint64   // Starting offset for lock
+	Length uint64   // Number of bytes to lock
+	PID    int32    // PID of process blocking our lock (F_GETLK only)
+
+	// "client_id is an additional mechanism for uniquely
+	// identifying the lock requester and is set to the nodename
+	// by the Linux v9fs client."
+	// https://github.com/chaos/diod/blob/master/protocol.md#lock---acquire-or-release-a-posix-record-lock
+	Client string // Client id -- but technically can be anything.
+}
+
+// decode implements encoder.decode.
+func (t *tlock) decode(b *buffer) {
+	t.fid = b.ReadFID()
+	t.Type = LockType(b.Read8())
+	t.Flags = b.Read32()
+	t.Start = b.Read64()
+	t.Length = b.Read64()
+	t.PID = int32(b.Read32())
+	t.Client = b.ReadString()
+}
+
+// encode implements encoder.encode.
+func (t *tlock) encode(b *buffer) {
+	b.WriteFID(t.fid)
+	b.Write8(uint8(t.Type))
+	b.Write32(t.Flags)
+	b.Write64(t.Start)
+	b.Write64(t.Length)
+	b.Write32(uint32(t.PID))
+	b.WriteString(t.Client)
+}
+
+// typ implements message.typ.
+func (*tlock) typ() msgType {
+	return msgTlock
+}
+
+// String implements fmt.Stringer.
+func (t *tlock) String() string {
+	return fmt.Sprintf("Tlock{Type: %s, Flags: %#x, Start: %d, Length: %d, PID: %d, Client: %s}", t.Type.String(), t.Flags, t.Start, t.Length, t.PID, t.Client)
+}
+
+// rversion is a version response.
+type rlock struct {
+	Status uint8
+}
+
+// decode implements encoder.decode.
+func (r *rlock) decode(b *buffer) {
+	r.Status = b.Read8()
+}
+
+// encode implements encoder.encode.
+func (r *rlock) encode(b *buffer) {
+	b.Write8(r.Status)
```

Comment:
Can we just say `0` (`lockOK`) here? Like, really ignore it?

Reply:
I think it's best to try. There were real world cases where failing to lock led to some issues.

Reply:
I was afraid that'd be the case. Now looking back at the lock failures in https://github.com/hugelgupf/p9/pull/52#discussion_r953069332 - we'd need some way to figure out the cause. I will happily try whatever, but I'm not yet familiar enough with the codebase to get further.